### PR TITLE
DB-11189 Avoid base table lookups for index updates (with DB-10765)

### PR DESCRIPTION
**Pull request by dgomezferro**
## Short Description
This change includes old values in the update KVPair so that we can delete the right index entry without performing a (potentially expensive) baseTable lookup for each index

## Long Description
During an UPDATE operation we generate an ExecRow with the old and the new row values ( old0, old1,... , new0, new1, ...). When we write this into the WritePipeline we only include the new values though, this is what the IndexWriteHandler receives. 

If we have an index, at this point we usually have to delete the old index entry and create a new one with the new values. To create a delete right now we perform a local lookup using the updated row's rowkey and requesting all columns the index has indexed. With that we create the delete and with the incoming new values we create the new entry.

We do this for every index, each one will perform its own local lookup with its own indexed columns. The lookup is local (not remote) because we are intercepting the write into the baseTable, so the data is right there, but it still can be expensive if we have to scan over a lot of MVCC history for instance (see DB-11189)

The proposed fix includes the old values in the write payload, encoding them as more columns ( new0, new1, ... , old0, old1, ...) The WritePipeline will then take the columns it need for each purpose:
- (old0, old1...) for the delete
- (new0, new1, ...) for the write

The drawback is that for tables with no indices we are adding a bit of overhead (serializing old values). We could be smarter about this. Longer term I think we should redo the whole index framework so that it's driven from the client.

### CODE REVIEW

**Comment by dgomezferro** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

File: pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/WriteHandler.java [hunk 1]
```diff
@@ -14,8 +14,13 @@
 
 package com.splicemachine.pipeline.writehandler;
 
+import com.splicemachine.encoding.MultiFieldDecoder;
 import com.splicemachine.kvpair.KVPair;
 import com.splicemachine.pipeline.context.WriteContext;
+import com.splicemachine.storage.EntryDecoder;
+import com.splicemachine.storage.index.BitIndex;
+import com.splicemachine.storage.index.BitIndexing;
+import com.splicemachine.utils.ByteSlice;
```
Comment (dgomezferro) [inline pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/WriteHandler.java:23]: Cleanup

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2762/

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java [hunk 2]
```diff
@@ -477,11 +483,13 @@ public boolean areIndexKeysModified(KVPair mutation, BitSet indexedColumns) {
...
-            if (indexedColumns.get(i))
-                return true;
+        BitSetIterator iterator = nonPkIndexedCols.iterator();
+        int nextBit;
+        while((nextBit = iterator.nextSetBit()) != -1) {
+            if (!updateIndex.isSet(nextBit))
+                return false;
         }
-        return false;
+        return true;
```
Comment (ascend1) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java:492]: It looks like parameter `indexedColumns` is not needed after the change.

File: splice_machine/src/main/java/com/splicemachine/derby/stream/output/update/UpdatePipelineWriter.java [hunk 3]
```diff
@@ -95,10 +96,12 @@ public void open(TriggerHandler triggerHandler, SpliceOperation operation, boole
...
         for(int i=0;i<columnOrdering.length;++i)
             kdvds[i]=LazyDataValueFactory.getLazyNull(formatIds[columnOrdering[i]]);
-        colPositionMap=new int[heapList.size()];
+        colPositionMap=new int[heapList.size()*2];
         // Map Column Positions for encoding
-        for(int i=heapList.anySetBit(), pos=heapList.getNumBitsSet();i!=-1;i=heapList.anySetBit(i),pos++)
-            colPositionMap[i]=pos;
+        for(int i=heapList.anySetBit(), pos=heapList.getNumBitsSet();i!=-1;i=heapList.anySetBit(i),pos++) {
+            colPositionMap[i] = pos;
+            colPositionMap[i+heapList.size()] = pos - heapList.getNumBitsSet();
```
Comment (arnaud-splice) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/output/update/UpdatePipelineWriter.java:103]: Can you add a comment about this part? The map being twice as long as before to store old and new values

**Review by arnaud-splice — APPROVED** [review]

**Review by martinrupp — CHANGES_REQUESTED** [review]
see comments on https://github.com/splicemachine/spliceengine/pull/5384

File: pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java [hunk 4]
```diff
@@ -9,9 +9,36 @@
 import com.splicemachine.storage.index.BitIndexing;
 import com.splicemachine.utils.ByteSlice;
 
-public abstract class AbstractWriteHandler implements WriteHandler {
+public class UpdateUtils {
 
-    protected KVPair getBaseUpdateMutation(KVPair mutation) {
+    public static KVPair deleteFromUpdate(KVPair mutation) {
```
Comment (ascend1) [inline pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java:14]: I think it makes sense to have a comment about the content of mutation passed in to this function. I feel the logic in this function that takes half of the fields can be confusing to someone who doesn't have the background of this problem. Just a suggestion.

**Review by ascend1 — APPROVED** [review]
I didn't get what problem the commit "DB-11189 Fix constraints and Mem platform" fixes. Otherwise it looks good.

File: pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        rowDecoder.set(mutation.getValue());
+        BitIndex index = rowDecoder.getCurrentIndex();
+        MultiFieldDecoder fieldDecoder = rowDecoder.getEntryDecoder();
+        int offset = fieldDecoder.offset();
+        for(int i=index.nextSetBit(0), c=0; c<index.cardinality()/2; i=index.nextSetBit(i+1), c++){
+            rowDecoder.seekForward(fieldDecoder, i);
+        }
+        int end = fieldDecoder.offset();
+        fieldDecoder.seek(offset);
+        byte[] data = fieldDecoder.slice(end - offset - 1); // -1 to remove the NULL separator
```
Comment (martinrupp) [inline pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java:52]: (copied from https://github.com/splicemachine/spliceengine/pull/5384)

```
int offset = fieldDecoder.offset();
int end = fieldDecoder.offset();
byte[] data = fieldDecoder.slice(end - offset - 1);
```
this looks like
`byte[] data = fieldDecoder.slice( - 1);`
????
am i missing something?

why is this so different from line 23?
`byte[] data = fieldDecoder.slice(fieldDecoder.length() - fieldDecoder.offset());`

Apart from this, this function looks like a copy&paste of the function deleteFromUpdate, except for KVPair.Type.UPDATE

**Review by martinrupp — CHANGES_REQUESTED** [review]

File: pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java (hunk 5 above)
Comment (martinrupp) [reply pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java:52]: ah ok `rowDecoder.seekForward(fieldDecoder, i);` would change this. But why is that code so different from deleteFromUpdate?

**Review by carolp-503 — APPROVED** [review]
+1

**Comment by dgomezferro** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

File: pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java (hunk 5 above)
Comment (dgomezferro) [reply pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java:52]: In order to decode the value we need to skip over one field at a time. In general fields are separated by a null value `0x00` but depending some types of fields (scalar, float & double) can contain null bytes, and the type information is encoded in the BitIndex structure.
So if we know we have 4 fields, we skip over 2 fields and get that offset. If we want the second half we take everything from that point till the end, if we want the first half remember that position, we `seek` the decoder back to the beginning and take everything from the beginning to the middle position.
I renamed some variables to hopefully make it more understandable.

File: pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java (hunk 4 above)
Comment (dgomezferro) [reply pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java:14]: I added that explanation to `extractWriteFromUpdate` after I refactored `deleteFromUpdate` and `getBaseUpdateMutation` as @martinrupp suggested.

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java (hunk 2 above)
Comment (dgomezferro) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java:492]: Good catch, removed.

File: splice_machine/src/main/java/com/splicemachine/derby/stream/output/update/UpdatePipelineWriter.java (hunk 3 above)
Comment (dgomezferro) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/output/update/UpdatePipelineWriter.java:103]: I added a short comment and a reference to `UpdateUtils.extractWriteFromUpdate`

**Comment by dgomezferro** [comment]
> I didn't get what problem the commit "DB-11189 Fix constraints and Mem platform" fixes. Otherwise it looks good.

@ascend1 the constraint fix has to do with exiting early in `UpdateNode.getUpdateReadMap`. Since I was including all columns on the select (if we update the PK) I thought we shouldn't keep checking constraints etc for more columns to include, but there were some side effects that were needed (for instance `baseTable.getAllRelevantConstraints` updates `relevantConstraints` )

The Mem platform fix was related to DB-10765, the `get()` method in MPartition (used when deleting an index entry) wasn't applying filters correctly. If it needed to merge multiple KVs to generate the result it would return the first KV. For instance if we are asking for columns {1, 3} and there's two KVs one with columns {0, 1} and another with columns {0,1,2,3} instead of taking column 0 from the first and column 3 from the second to materialize the result correctly, it would return the first KV with values {0, 1} so we'd delete the wrong index entry (in practice we wouldn't delete the index entry). That was behind this jenkins failure http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2748/

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by dgomezferro** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2766/

**Comment by dgomezferro** [comment]
jenkins please test branch @cdh6.3.0

**Review by martinrupp — APPROVED** [review]

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**